tests/eve_tests/unit_tests/test_datamodels.py

```diff
@@ -0,0 +1,842 @@
+def test_root_validators(model_class: Type[datamodels.DataModelTp]):
+    model_class(int_value=0, float_value=1.1, str_value="")
+    with pytest.raises(ValueError, match="float_value"):
+        model_class(int_value=1, float_value=1.0, str_value="")
+
+
+def test_root_validators_in_subclasses():
+    class Model(ModelWithRootValidators):
+        @datamodels.root_validator
+        def _root_validator(cls, instance):
+            assert cls.class_counter == 2
+            cls.class_counter += 10
+
+        @datamodels.root_validator
+        def _another_root_validator(cls, instance):
+            assert cls.class_counter == 12
+            cls.class_counter += 10
+
+        @datamodels.root_validator
+        def _final_root_validator(cls, instance):
+            assert cls.class_counter == 22
+            if str(instance.int_value) == instance.str_value:
+                raise ValueError("'int_value' and 'str_value' must be different")
+
+    with pytest.raises(ValueError, match="str_value"):
+        Model(int_value=1, float_value=0.0, str_value="1")
+    with pytest.raises(ValueError, match="float_value"):
+        Model(int_value=1, float_value=1.0, str_value="1")
+    with pytest.raises(ValueError, match="float_value"):
+        Model(int_value=1, float_value=1.0, str_value="1")
+
+
+# Test field options
+def test_field_init():
+    @datamodels.datamodel
+    class Model:
+        hidden_value: int = datamodels.field(init=False)
+        value: int
+        hidden_value_with_default: int = datamodels.field(default=10, init=False)
+
+    model = Model(value=1)
+    assert model.value == 1
+    assert model.hidden_value_with_default == 10
+
+    model.hidden_value = -1
+    assert model.hidden_value == -1
+
+
+def test_field_metadata():
+    @datamodels.datamodel
+    class Model:
+        value: int = datamodels.field(metadata={"my_metadata": "META"})
+
+    assert isinstance(Model.__datamodel_fields__.value.metadata, types.MappingProxyType)
+    assert Model.__datamodel_fields__.value.metadata["my_metadata"] == "META"
+
+
+# Test datamodel options
+def test_frozen():
+    import attr
+
+    @datamodels.datamodel(frozen=True)
+    class FrozenModel:
+        value: Any = None
+
+    assert FrozenModel.__datamodel_params__.frozen is True
+    with pytest.raises(attr.exceptions.FrozenInstanceError):
+        FrozenModel().value = 1
+
+    class FrozenModel2(datamodels.DataModel, frozen=True):
+        value: Any = None
+
+    assert FrozenModel2.__datamodel_params__.frozen is True
+    with pytest.raises(attr.exceptions.FrozenInstanceError):
+        FrozenModel2().value = 1
+
+
+def test_hash():
+    string_value = "this is a really long string to avoid string interning 1234567890 +:?.!"
+
+    # Safe mutable case: no hash
+    @datamodels.datamodel
+    class MutableModel:
+        value: Any = None
+
+    with pytest.raises(TypeError, match="unhashable type"):
+        hash(MutableModel(value=string_value))
```

Comment:
Seems useful to document the motivation behind this behavior. This effectively prevents the creating of dictionaries with mutable models as keys. This case however should be prevented as attaching information to models which are mutable might be unsafe. When the model is mutated the attached information might not be valid anymore.

Reply:
This is the default Python / attrs / dataclasses behavior. I will add a link for reference.